components/dashboard/src/settings/PersonalAccessTokens.tsx

```diff
@@ -452,33 +606,85 @@ function ListAccessTokensView() {
-                        No Personal Access Tokens (PAT)
-                    </h3>
-                    <p className="text-center pb-6 text-gray-500 text-base w-96">
-                        Generate a personal access token (PAT) for applications that need access to the Gitpod API.{" "}
-                    </p>
-                    <Link to={settingsPathPersonalAccessTokenCreate}>
-                        <button>New Personal Access Token</button>
-                    </Link>
-                </div>
+            {loading ? (
+                <SpinnerLoader content="fetch data" />
             ) : (
                 <>
-                    <div className="px-6 py-3 flex justify-between space-x-2 text-sm text-gray-400 mb-2 bg-gray-100 dark:bg-gray-800 rounded-xl">
-                        <h2 className="w-3/12">Token Name</h2>
-                        <h2 className="w-3/12">Permissions</h2>
-                        <h2 className="w-3/12">Expires</h2>
-                        <div className="w-3/12"></div>
-                    </div>
-                    {tokens.map((t: PersonalAccessToken) => {
-                        return <TokenEntry token={t} onDelete={handleDeleteToken} />;
-                    })}
+                    {tokens.length === 0 ? (
+                        <div className="bg-gray-100 dark:bg-gray-800 rounded-xl w-full py-28 flex flex-col items-center">
+                            <h3 className="text-center pb-3 text-gray-500 dark:text-gray-400">
+                                No Personal Access Tokens (PAT)
+                            </h3>
+                            <p className="text-center pb-6 text-gray-500 text-base w-96">
+                                Generate a personal access token (PAT) for applications that need access to the Gitpod
+                                API.{" "}
+                            </p>
+                            <Link to={settingsPathPersonalAccessTokenCreate}>
+                                <button>New Personal Access Token</button>
+                            </Link>
+                        </div>
+                    ) : (
+                        <>
+                            <div className="px-3 py-3 flex justify-between space-x-2 text-sm text-gray-400 mb-2 bg-gray-100 dark:bg-gray-800 rounded-xl">
+                                <h2 className="w-4/12">Token Name</h2>
+                                <h2 className="w-4/12">Permissions</h2>
+                                <h2 className="w-3/12">Expires</h2>
+                                <div className="w-1/12"></div>
+                            </div>
+                            {tokens.map((t: PersonalAccessToken) => (
+                                <TokenEntry
+                                    key={t.id}
+                                    token={t}
+                                    menuEntries={[
+                                        {
+                                            title: "Edit",
```

Comment:
**issue**: Not the best line to comment this, but when you _Edit_ a token and add or remove permissions, the expiration date changes back to today. ❗

Reply:
It can be reproduced.

But it's weird that [from the code](https://github.com/gitpod-io/gitpod/blob/7ffe8cccd33dacdca0f385ce8efe0245f9459707/components/dashboard/src/settings/PersonalAccessTokens.tsx#L269-L276), we don't send `expirationTime`, but devTools shows we have send it (maybe a default value of `protobuf` `Timestamp`)

<img width="769" alt="image" src="https://user-images.githubusercontent.com/20944364/204807712-9b3bd91c-4830-4d32-8a19-496c914f3094.png">


And server should not update expirationDate too.

cc @easyCZ

Reply:
For the fine grained update, there's a second field which you can supply - `updateMask`. It's a list of of the `token` fields which should be updated.

It may make sense to link the update form "changes" to those fields and send them to ensure its updated. I'll investigate and can look into fixing the backend, it should never send back `internal` error code anyway.

Reply:
I've added a couple of fixes to API validation and Not Found handling in the following:
* https://github.com/gitpod-io/gitpod/pull/15075
* https://github.com/gitpod-io/gitpod/pull/15072

That should at least give us more info about the error.

Reply:
Okay, with the two fixes the error now becomes 
```
code: "not_found"
message: "Personal Access Token with ID 35ebd6ef-57d1-44fe-9309-b91b16d8565a for User a86d46b2-5f89-4be3-a3b5-fa676edaadb7 does not exist"
```

I won't get around to investigate/fix today, so all yours if you want. Otherwise I'll look into it in the morning (CET)

Reply:
@easyCZ we took a look with @mustard-mh and @iQQBot and for some reason the expirationTime has onUpdate :thinking: no idea where this is coming from 
![image](https://user-images.githubusercontent.com/25115070/204886373-87176d97-eee1-453b-8f4d-c46023915eb2.png)

Reply:
We found the root cause, it's because our DB is using `explicit_defaults_for_timestamp=off`
There's a comment about it here https://github.com/gitpod-io/gitpod/blob/7ffe8cccd33dacdca0f385ce8efe0245f9459707/components/gitpod-db/src/typeorm/transformer.ts#L28-L33
So we need a new migration setting default value on the column

Reply:
Looks like the original migration put the `on update` on the wrong column. The on-update should be on set on _lastModified

Reply:
Great find! I'll add a migration tomorrow.

Reply:
Migration PR: https://github.com/gitpod-io/gitpod/pull/15093

Reply:
The update issue, where the expiration would be reset is now fixed in staging.

Reply:
The remaining issue is that a no-update returns a Not Found from the API. Tracked here https://github.com/gitpod-io/gitpod/issues/15105